sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncActivityViewModel.kt

```diff
@@ -50,98 +54,186 @@ class SyncActivityViewModel @Inject constructor(
+                }
             }
         }
     }
 
-    private suspend fun updateViewState() {
-        val qrBitmap = withContext(dispatchers.io()) {
-            val recoveryCode = syncRepository.getRecoveryCode() ?: return@withContext null
-            qrEncoder.encodeAsBitmap(recoveryCode, R.dimen.qrSizeLarge, R.dimen.qrSizeLarge)
+    private suspend fun fetchRemoteDevices() {
+        withContext(dispatchers.io()) {
+            val result = syncRepository.getConnectedDevices()
+            if (result is Success) {
+                val newState = viewState.value.hideDeviceListItemLoading().setDevices(result.data.map { SyncedDevice(it) })
+                viewState.emit(newState)
+            } else {
+                viewState.emit(viewState.value.hideDeviceListItemLoading())
+            }
         }
-        viewState.emit(
-            viewState.value.copy(
-                isDeviceSyncEnabled = syncRepository.isSignedIn(),
-                showAccount = syncRepository.isSignedIn(),
-                loginQRCode = qrBitmap,
-            ),
-        )
     }
 
-    fun onTurnOffSyncConfirmed() {
+    fun onTurnOffSyncConfirmed(connectedDevice: ConnectedDevice) {
         viewModelScope.launch(dispatchers.io()) {
-            viewState.emit(viewState.value.copy(showAccount = false))
-            val deviceId = syncRepository.getThisConnectedDevice().deviceId
-            when (syncRepository.logout(deviceId)) {
-                is Error -> {
-                    updateViewState()
-                }
-                is Success -> {
-                    updateViewState()
-                }
-            }
+            viewState.emit(viewState.value.hideAccount())
+            syncRepository.logout(connectedDevice.deviceId)
+            hideOrShowAccountDetails()
         }
     }
 
     fun onTurnOffSyncCancelled() {
         viewModelScope.launch {
-            viewState.emit(viewState.value.copy(isDeviceSyncEnabled = true))
+            hideOrShowAccountDetails()
         }
     }
 
     fun onDeleteAccountClicked() {
         viewModelScope.launch {
-            viewState.emit(viewState.value.copy(isDeviceSyncEnabled = false))
+            viewState.emit(viewState.value.toggle(false))
             command.send(AskDeleteAccount)
         }
     }
 
     fun onDeleteAccountConfirmed() {
         viewModelScope.launch(dispatchers.io()) {
-            viewState.emit(viewState.value.copy(showAccount = false))
-            when (syncRepository.deleteAccount()) {
-                is Error -> {
-                    Timber.i("deleteAccount failed")
-                    updateViewState()
-                }
-                is Success -> {
-                    Timber.i("deleteAccount success")
-                    updateViewState()
-                }
-            }
+            viewState.emit(viewState.value.hideAccount())
+            syncRepository.deleteAccount()
+            hideOrShowAccountDetails()
```

Comment:
why are you emitting two view states after each other? is this really necessary?

Reply:
here we emit hideAccount for update the UI without waiting for the backend response. If it fails, the UI will show the account again. Just an optimistic approach